src/js/trove/parse-pyret.js

```diff
@@ -1344,31 +1345,56 @@
         if (countParses === 1) {
           var ast = grammar.constructUniqueParse(parsed);
           //          console.log(ast.toString());
-          return translate(ast, fileName);
+          return RUNTIME.ffi.makeRight(translate(ast, fileName));
         } else {
           var asts = grammar.constructAllParses(parsed);
           throw "Non-unique parse";
           for (var i = 0; i < asts.length; i++) {
             //console.log("Parse " + i + ": " + asts[i].toString());
             //            console.log(("" + asts[i]) === ("" + asts2[i]));
           }
-          return translate(ast, fileName);
+          return RUNTIME.ffi.makeRight(translate(ast, fileName));
         }
       } catch(e) {
-        // console.error("Fatal error in parsing: ", e);
-        throw e;
+        if (RUNTIME.isPyretException(e)) {
+          return RUNTIME.ffi.makeLeft(RUNTIME.makeObject({
```

Comment:
The other pattern we have for this is `execThunk`, which is exposed as `run-task`.

In general, I don't know what cases this would throw a _non_ Pyret error.

Also, in general I think using `catch` like this isn't safe with the stack, because the handling code isn't added to the frame.  I think in this case the parser is assumed to be flat-ish (since all it does is call constructors), so the `catch` works.

I'm not bothered by the unwrapping below – that just seems like how the two versions of this has to work.  The exception-throwing version unwrapping an `Either` is a common pattern, right?  It could be done the opposite way, where `maybe-surface-parse` catches the exception and wraps, instead of `parseDataRaw` doing it, but I don't see a big difference.

I think I'm OK with this, since the construction part of parsing has to be flat anyway and has all sorts of issues with stack-unsafe constructors as it is.  So unless someone points out something I'm misunderstanding, I'm happy merging this.